include/sparrow/c_interface.hpp

```diff
@@ -0,0 +1,437 @@
+
+        // Release callback
+        void (*release)(struct ArrowSchema*) = nullptr;
+        // Opaque producer-specific data
+        void* private_data = nullptr;
+    };
+
+    struct ArrowArray
+    {
+        // Array data description
+        int64_t length = 0;
+        int64_t null_count = 0;
+        int64_t offset = 0;
+        int64_t n_buffers = 0;
+        int64_t n_children = 0;
+        const void** buffers = nullptr;
+        struct ArrowArray** children = nullptr;
+        struct ArrowArray* dictionary = nullptr;
+
+        // Release callback
+        void (*release)(struct ArrowArray*) = nullptr;
+        // Opaque producer-specific data
+        void* private_data = nullptr;
+    };
+
+}  // extern "C"
+
+namespace sparrow
+{
+    template <class T>
+    std::vector<T*> to_raw_ptr_vec(const std::vector<std::unique_ptr<T>>& vec)
+    {
+        std::vector<T*> raw_ptr_vec;
+        std::ranges::transform(
+            vec,
+            std::back_inserter(raw_ptr_vec),
+            [](const auto& child)
+            {
+                return child.get();
+            }
+        );
+        return raw_ptr_vec;
+    }
+
+    template <class T>
+    std::vector<T*> to_raw_ptr_vec(std::vector<std::vector<T>>& vectors)
+    {
+        std::vector<T*> raw_ptr_vec;
+        std::ranges::transform(
+            vectors,
+            std::back_inserter(raw_ptr_vec),
+            [](std::vector<T>& vector) -> T*
+            {
+                return vector.data();
+            }
+        );
+        return raw_ptr_vec;
+    }
+
+    /**
+     * Struct representing private data for ArrowArray.
+     *
+     * This struct holds the private data for ArrowArray, including buffers, children, and dictionary.
+     * It is used in the Sparrow library.
+     *
+     * @tparam BufferType The type of the buffers.
+     * @tparam Allocator An allocator for BufferType.
+     */
+    template <class BufferType, template <typename> class Allocator = std::allocator>
+        requires sparrow::allocator<Allocator<BufferType>>
+    struct ArrowArrayPrivateData
+    {
+        explicit ArrowArrayPrivateData(
+            std::vector<std::unique_ptr<ArrowArray>> children,
+            std::unique_ptr<ArrowArray> dictionary,
+            size_t n_buffers,
+            size_t buffer_length
+        )
+            : buffers_(n_buffers, std::vector<BufferType, Allocator<BufferType>>(buffer_length))
+            , buffers_raw_ptr_vec_(to_raw_ptr_vec(buffers_))
+            , children_(std::move(children))
+            , children_raw_ptr_vec_(to_raw_ptr_vec(children_))
+            , dictionary_(std::move(dictionary))
+        {
+        }
+
+        using BufferAllocator = Allocator<BufferType>;
+        using BuffersAllocator = Allocator<std::vector<BufferType, BufferAllocator>>;
+
+        any_allocator<BufferType> buffer_allocator_ = BufferAllocator();
+
+        std::vector<std::vector<BufferType, Allocator<BufferType>>> buffers_;
```

Comment:
- `BufferAllocator` (or even better, `buffer_allocator`), can be used instead of `Allocator<BufferType>`
- Why `std::vector<std::vector<BufferType, [...]>>` instead of `std::vector<buffer<BufferType, [... ]>>` ? This would allow to have a single structure for the use case where the Arrow Array owns its memory, and the use case where it is used as a view to another structure suche as sparrow::array.